### PR TITLE
Class16 week4

### PR DESCRIPTION
**Pull request by AllouhM**

### CODE REVIEW

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  handleDescriptionUpdate = e => {
+    this.setState({
+      description: e.target.value
+    });
+  };
+
+  handleDeadlineUpdate = date => {
+    this.setState({ deadline: date });
+  };
+  handleSubmit = (todoId, event) => {
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/components/UpdateTodo.jsx]: This method have to be in the store as an action.

So you will only pass from here the state, for making things organized, means: store have the actions, component have the state.

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    event.preventDefault();
+    const { TodosStore } = this.props;
+    TodosStore.handleUpdate(
+      todoId,
+      this.state.description,
+      this.state.deadline
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/components/UpdateTodo.jsx]: Excellent use of a React state with Mobx

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        </div>
+        <React.Fragment>
+          <Form
+            stateData={this.state}
+            todo={todo}
+            handleSubmit={this.handleSubmit}
+            handleDescriptionUpdate={this.handleDescriptionUpdate}
+            handleDeadlineUpdate={this.handleDeadlineUpdate}
+          />
+        </React.Fragment>
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/components/UpdateTodo.jsx]: You are having a `React.Fragment` on a Component have been wrapped in a `div` you may only change the `div` to be the `React.Fragment` and remove this one, or just remove this 🙂

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import { observable, computed, action } from "mobx";
+import Data from "../Data.json";
+import { configure } from "mobx";
+configure({ enforceActions: true });
+
+class TodosStore {
+  @observable
+  edit = false;
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/stores/TodosStore.jsx]: you have this observable with no use.

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  @observable
+  Data = Data;
+
+  @action
+  handleCheck = todoId => {
+    const newData = this.Data.map(todo => {
+      if (todo.id === todoId) {
+        return { ...todo, done: !todo.done };
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/stores/TodosStore.jsx]: Great use of the spread operator, but you can leave it in MobX, in state it's very important to not mutate it and that's why I was using it always and others around.

if you chosed to use it here you may use it everywhere or no where, in order to not confuse yourself where you have to use it, or others they have to read es6 or es5

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      }
+      return todo;
+    });
+    this.Data = newData;
+  };
+  @action
+  handleDelete = todoId => {
+    const filteredTodos = this.Data.filter(todo => {
+      return todo.id !== todoId;
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/stores/TodosStore.jsx]: as I told before this is okay in MobX, while in the state this shouldn't be the case.

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  };
+
+  @action
+  handleUpdate = (todoId, description, deadline) => {
+    const newData = this.Data.map(todo => {
+      if (todo.id === todoId) {
+        return {
+          ...todo,
+          description: description || todo.description,
+          deadline: deadline || todo.deadline
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/stores/TodosStore.jsx]: 🎉

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+      return todo;
+    });
+    this.Data = newData;
+  };
+
+  @computed
+  get remainingTodos() {
+    const remaining = this.Data.filter(todo => {
+      return todo.done !== true;
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/stores/TodosStore.jsx]: shorthand 🙂 
```suggestion
      return !todo.done;
```

File: homework-submission/week4-class16/src/components/Functionality.jsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from "react";
+import { observer, inject } from "mobx-react";
+@inject("TodosStore")
+@observer
+class Functionality extends Component {
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/components/Functionality.jsx]: I am not sure till now what this component is doing 😕

File: homework-submission/week4-class16/src/components/Functionality.jsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  render() {
+    const { todo } = this.props;
+    const { TodosStore } = this.props;
+
+    return (
+      <div className="function-container">
+        <label className="check_box">
+          <input
+            type="checkbox"
+            onClick={() => TodosStore.handleCheck(todo.id)}
```
Comment (Jawhar-B) [inline homework-submission/week4-class16/src/components/Functionality.jsx]: Aha! it's checking done, and deleting, But what if this component have around 200 lines? which is possible...

**Comment by AllouhM** [comment]
Thank you very much for the time you spend in leaving your useful feedback.

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    );
+
+    const i = TodosStore.Data.findIndex(todo => todo.id === todoId);
+    TodosStore.Data[i].edit = !TodosStore.Data[i].edit;
+  };
+
+  render() {
+    const { todo } = this.props;
+    const { TodosStore } = this.props;
+    console.log(TodosStore);
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/UpdateTodo.jsx]: amended in week 5

File: homework-submission/week4-class16/src/App.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from "react";
+import { observer, inject } from "mobx-react";
+import Todolist from "./components/TodoList";
+import "./App.css";
+import "./todos.css";
+const image =
+  "https://i.postimg.cc/gjLMyGks/to-do-list-png-the-power-of-a-to-do-list-imodelafrica-1024.png";
+
+@inject("TodosStore")
+@observer
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/App.js]: amended in week 5

File: homework-submission/week4-class16/src/stores/TodosStore.jsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
+import { observable, computed, action } from "mobx";
+import Data from "../Data.json";
+import { configure } from "mobx";
+configure({ enforceActions: true });
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/stores/TodosStore.jsx]: amended in week 5

File: homework-submission/week4-class16/src/components/TodoList.jsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+        </li>
+      ) : (
+        todos
+      );
+    const leftTodos =
+      TodosStore.remainingTodos === 1
+        ? `You still have 
+          ${TodosStore.remainingTodos} todo needs action`
+        : `You still have 
+          ${TodosStore.remainingTodos} todos need action`;
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/TodoList.jsx]: it is for singular or plural

File: homework-submission/week4-class16/src/stores/TodosStore.jsx (hunk 8 above)
Comment (AllouhM) [reply homework-submission/week4-class16/src/stores/TodosStore.jsx]: amended in week 5

File: homework-submission/week4-class16/src/components/Functionality.jsx (hunk 10 above)
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/Functionality.jsx]: I didn't get your point here

File: homework-submission/week4-class16/src/components/TodoList.jsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    const { TodosStore } = this.props;
+    const list = TodosStore.Data;
+    const todos = list.map(todo => {
+      return <Todo key={todo.id} todo={todo} />;
+    });
+    const rendering =
+      list.length === 0 ? (
+        <li className="zero_todo">
+          {" "}
+          The List IS Empty. Please add new todo....
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/TodoList.jsx]: it is added automatically from prettier

File: homework-submission/week4-class16/src/components/TodoList.jsx [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+@observer
+class TodoList extends Component {
+  render() {
+    const { TodosStore } = this.props;
+    const list = TodosStore.Data;
+    const todos = list.map(todo => {
+      return <Todo key={todo.id} todo={todo} />;
+    });
+    const rendering =
+      list.length === 0 ? (
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/TodoList.jsx]: updated in week 5

File: homework-submission/week4-class16/src/components/Todo.jsx [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import Functionality from "./Functionality";
+import Moment from "react-moment";
+import UpdateTodo from "./UpdateTodo";
+// import { observer } from "mobx-react";
+
+// @observer
```
[the remark this answers is not in the document]
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/Todo.jsx]: updated in week 5

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx (hunk 1 above)
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/UpdateTodo.jsx]: you mean handleSubmit, right? actually, this function calls in the next line handleUpdate function which indeed an action in the store.

File: homework-submission/week4-class16/src/components/UpdateTodo.jsx (hunk 3 above)
Comment (AllouhM) [reply homework-submission/week4-class16/src/components/UpdateTodo.jsx]: updated in week 5

**Pull request closed without merging**